### PR TITLE
Infer user-defined enum classes by checking if the class is a subtype of ``enum.Enum``

### PR DESCRIPTION
**Pull request by mbyrnepr2**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|    | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description

Infer user-defined enum classes by checking if the class is a subtype of ``enum.Enum``.

Closes pylint-dev/pylint#8897

### CODE REVIEW

File: astroid/brain/brain_namedtuple_enum.py [hunk 1]
```diff
@@ -644,14 +636,7 @@ def _get_namedtuple_fields(node: nodes.Call) -> str:
 
 def _is_enum_subclass(cls: astroid.ClassDef) -> bool:
     """Return whether cls is a subclass of an Enum."""
-    try:
-        return any(
-            klass.name in ENUM_BASE_NAMES
-            and getattr(klass.root(), "name", None) == "enum"
-            for klass in cls.mro()
```
Comment (mbyrnepr2) [inline astroid/brain/brain_namedtuple_enum.py:651]: `cls.mro()` wasn't returning the `enum` ancestors in the case of the example on the original issue; hence the class never made became inferred.

**Review by Pierre-Sassoulas — COMMENTED** [review]
On top of the fix, it seems like a more sensible approach indeed, good job. Did we have tests for each hard coded subclasses from the removed constants ? (I want to make sure that intEnum are still handled correctly)

**Comment by mbyrnepr2** [comment]
> On top of the fix, it seems like a more sensible approach indeed, good job. Did we have tests for each hard coded subclasses from the removed constants ? (I want to make sure that intEnum are still handled correctly)

Yes for [IntEnum](https://github.com/pylint-dev/astroid/blob/main/tests/brain/test_enum.py#L112).
No for `IntFlag` - there's tests for it in [Pylint](https://github.com/pylint-dev/pylint/blob/main/tests/functional/i/implicit/implicit_flag_alias.py#L2) but makes sense to have one here in astroid with the others.

**Comment by mbyrnepr2** [comment]
This change to using `is_subtype_of` has opened the door to another issue as can be seen in the failing unittest. The failing test expects the class to be subscriptable and these classes currently don't have `__getitem__` methods. Not sure how straightforward that fix will be. I hope to dig into it soon.

**Comment by jacobtylerwalls** [comment]
Sorry if this test will be difficult to rewrite! Maybe an extra inference call in the setup?

**Comment by mbyrnepr2** [comment]
@jacobtylerwalls I've tried a few things to modify the test but no luck; when the program enters the enum brain module and into the `is_subtype_of` function we are getting a StopIteration (if I recall correctly) and the context path doesn't behave as expected. I've also had little time lately 😬.

File: astroid/brain/brain_namedtuple_enum.py (hunk 1 above)
Comment (jacobtylerwalls) [reply astroid/brain/brain_namedtuple_enum.py:660]: Looks like a slight performance improvement, too, based on the small sample size of profiling astroid.

**Review by jacobtylerwalls — APPROVED** [review]

**Pull request merged**